independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/PooledBufferAllocator.java

```diff
@@ -0,0 +1,107 @@
+package io.quarkus.bootstrap.runner;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A thread-safe buffer allocator that pools heap buffers.<br>
+ * It's currently implemented using a concurrent linked stack, but the users shouldn't rely on how it's implemented,
+ * as it's subject to change. Conversely, users can rely on the fact that it's thread-safe.
+ *
+ * Users which relies on its buffer API, but doesn't need anymore to pool buffers, can call {@link #disablePooling()}
+ * which will make it behave as if it was disabled from the start and release any buffer that was still pooled.
+ */
+public final class PooledBufferAllocator {
+
+    // This is a sentinel value used to forcibly disable pooling.
+    private static final Buffer EMPTY = new Buffer();
+
+    /**
+     * A buffer that contains the data of a resource.<br>
+     * It can be released back (once) to its allocator via {@link PooledBufferAllocator#release(Buffer)}.
+     */
+    public static final class Buffer {
+        private byte[] data;
+        private int length;
+        private Buffer next;
+
+        private Buffer(byte[] data, int length) {
+            assert data != null && length >= 0;
+            this.data = data;
+            this.length = length;
+        }
+
+        private Buffer() {
+        }
+
+        public byte[] array() {
+            return data;
+        }
+
+        public int length() {
+            return length;
+        }
+
+        /**
+         * It moves the data from this buffer to a new one, and returns the new one.
+         */
+        private Buffer moveDataToNewBufferOf(final int requiredCapacity) {
+            byte[] data = this.data;
+            // it's forbidden to modify EMPTY
+            if (this != EMPTY) {
```

Comment:
this should be unnecessary? EMPTY will always have data.lenght == 0. So the below check will fail, unless requiredCapacity happens to be zero to, which ironically would imply it's fine to reuse EMPTY.

Reply:
BTW I realize you're trying to avoid marking `data` and `lenght` as `final` but the above comment would be self-explanatory if you did simply mark them as final.. perhaps it's not worth it?

Reply:
I'm easing the life of the GC by having `data` able to be nulled out while it is "moved" to another node: this will save the GC to figure it out later and using the usual "GC store barrier"s to remove the reachability of `data` from the old node ref's field